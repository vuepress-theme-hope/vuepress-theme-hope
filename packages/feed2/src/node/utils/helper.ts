import {
  Logger,
  isLinkHttp,
  removeEndingSlash,
  removeLeadingSlash,
} from "vuepress-shared/node";

export const FEED_GENERATOR = "vuepress-plugin-feed2";

export const logger = new Logger(FEED_GENERATOR);

export const resolveUrl = (hostname: string, base = "", path = ""): string =>
  `${
    isLinkHttp(hostname)
      ? removeEndingSlash(hostname)
      : `https://${removeEndingSlash(hostname)}`
  }${base}${removeLeadingSlash(path)}`;

export const getImageMineType = (ext = ""): string =>
  `image/${
    ext === "jpg"
      ? "jpeg"
      : ext === "svg"
      ? "svg+xml"
      : ext === "jpeg" ||
        ext === "png" ||
        ext === "bmp" ||
        ext === "gif" ||
        ext === "webp"
      ? ext
      : ""
  }`;
