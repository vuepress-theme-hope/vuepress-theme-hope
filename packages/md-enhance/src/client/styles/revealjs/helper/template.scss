@mixin normalize() {
  font-weight: normal;

  p {
    line-height: 1.3;
  }

  /* Ensure certain elements are never larger than the slide itself */
  img,
  video,
  iframe {
    max-width: 95%;
    max-height: 95%;
  }

  strong,
  b {
    font-weight: bold;
  }

  em {
    font-style: italic;
  }

  ol,
  dl,
  ul {
    display: inline-block;
    margin: 0 0 0 1em;
    text-align: start;
  }

  ol {
    list-style-type: decimal;
  }

  ul ul,
  ul ol,
  ol ol,
  ol ul {
    display: block;
    margin-left: 40px;
  }

  // stylelint-disable-next-line  no-descending-specificity
  ul {
    list-style-type: disc;

    ul {
      list-style-type: square;

      // stylelint-disable selector-max-compound-selectors
      ul {
        list-style-type: circle;
      }
      // stylelint-enable selector-max-compound-selectors
    }
  }

  dt {
    font-weight: bold;
  }

  dd {
    margin-left: 40px;
  }

  blockquote {
    position: relative;

    display: block;

    width: 70%;
    padding: 5px;

    background: rgb(255 255 255 / 5%);
    box-shadow: 0 0 2px rgb(0 0 0 / 20%);

    font-style: italic;

    p {
      &:first-child,
      &:last-child {
        display: inline-block;
      }
    }
  }

  q {
    font-style: italic;
  }

  pre {
    position: relative;

    display: block;

    width: 90%;
    padding: 0;

    box-shadow: 0 5px 15px rgb(0 0 0 / 15%);

    font-size: 0.55em;
    line-height: 1.2em;
    text-align: left;
    word-wrap: break-word;
  }

  code {
    text-transform: none;
  }

  pre code {
    display: block;

    overflow: auto;

    max-height: 400px;
    padding: 5px;

    word-wrap: normal;
  }

  a {
    text-decoration: none;
    transition: color 0.15s ease;

    &:hover {
      border: none;
      text-shadow: none;
    }
  }

  table {
    margin: auto;
    border-spacing: 0;
    border-collapse: collapse;
  }

  th {
    font-weight: bold;
  }

  th,
  td {
    padding: 0.2em 0.5em;
    border-bottom: 1px solid;
    text-align: left;

    &[align="center"] {
      text-align: center;
    }

    &[align="right"] {
      text-align: right;
    }
  }

  // stylelint-disable selector-max-compound-selectors
  tbody tr:last-child {
    th,
    td {
      border-bottom: none;
    }
  }
  // stylelint-enable selector-max-compound-selectors

  sup {
    vertical-align: super;
    font-size: smaller;
  }

  sub {
    vertical-align: sub;
    font-size: smaller;
  }

  small {
    display: inline-block;
    vertical-align: top;
    font-size: 0.6em;
    line-height: 1.2em;

    * {
      vertical-align: top;
    }
  }
}

@mixin template() {
  .reveal-viewport {
    background-color: var(--r-background-color);
  }

  .reveal {
    @include normalize;
    color: var(--r-main-color);
    font-size: var(--r-main-font-size);
    font-family: var(--r-main-font);

    @media print {
      .backgrounds {
        background-color: var(--r-background-color);
      }
    }

    ::selection {
      background: var(--r-selection-background-color);
      color: var(--r-selection-color);
      text-shadow: none;
    }

    // stylelint-disable selector-max-compound-selectors
    .slides section {
      &,
      > section {
        font-weight: inherit;
        line-height: 1.3;
      }
    }
    // stylelint-enable selector-max-compound-selectors

    h1 {
      font-size: var(--r-heading1-size);
      text-shadow: var(--r-heading1-text-shadow);
    }

    h2 {
      font-size: var(--r-heading2-size);
    }

    h3 {
      font-size: var(--r-heading3-size);
    }

    h4 {
      font-size: var(--r-heading4-size);
    }

    p {
      margin: var(--r-block-margin) 0;
    }

    h1,
    h2,
    h3,
    h4,
    h5,
    h6 {
      margin: var(--r-heading-margin);

      color: var(--r-heading-color);

      font-weight: var(--r-heading-font-weight);
      font-family: var(--r-heading-font);
      line-height: var(--r-heading-line-height);
      letter-spacing: var(--r-heading-letter-spacing);
      text-transform: var(--r-heading-text-transform);
      text-shadow: var(--r-heading-text-shadow);
      word-wrap: break-word;

      &:last-child {
        margin-bottom: 0;
      }
    }

    blockquote {
      margin: var(--r-block-margin) auto;
    }

    pre {
      margin: var(--r-block-margin) auto;
      font-family: var(--r-code-font);
    }

    code {
      font-family: var(--r-code-font);
    }

    img {
      margin: var(--r-block-margin) 0;
    }

    a {
      color: var(--r-link-color);

      &:hover {
        color: var(--r-link-color-hover);
      }
    }

    .roll span::after {
      background: var(--r-link-color-dark);
      color: #fff;
    }

    .r-frame {
      border: 4px solid var(--r-main-color);
      box-shadow: 0 0 10px rgb(0 0 0 / 15%);
    }

    a .r-frame {
      transition: all 0.15s linear;
    }

    a:hover .r-frame {
      border-color: var(--r-link-color);
      box-shadow: 0 0 20px rgb(0 0 0 / 55%);
    }

    .controls {
      color: var(--r-link-color);
    }

    .progress {
      background: rgb(0 0 0 / 20%);
      color: var(--r-link-color);
    }

    table.hljs-ln {
      display: table;
      width: 100%;

      tr {
        border: none;

        &:nth-child(2n) {
          background: inherit !important;
        }
      }

      td {
        border: none;
      }
    }
  }
}
