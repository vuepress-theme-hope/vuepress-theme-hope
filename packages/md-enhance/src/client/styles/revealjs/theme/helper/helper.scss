@use "sass:map";
@use "sass:meta";
@use "vars";

$default-variables: meta.module-variables("vars");

@function circle-gradient($outer, $inner, $type: circle) {
  @return radial-gradient(
    center,
    $type farthest-corner,
    $inner 0%,
    $outer 100%
  );
}

@mixin lightmode-text-color($color) {
  section.has-light-background {
    &,
    h1,
    h2,
    h3,
    h4,
    h5,
    h6 {
      color: $color;
    }
  }
}

@mixin darkmode-text-color($color) {
  section.has-dark-background {
    &,
    h1,
    h2,
    h3,
    h4,
    h5,
    h6 {
      color: $color;
    }
  }
}

@mixin theme($theme, $theme-variables) {
  $variables: map.deep-merge($default-variables, $theme-variables);
  $viewport-color: map.get($variables, "viewport-color");
  $background-color: map.get($variables, "background-color");
  $main-font: map.get($variables, "main-font");
  $main-font-size: map.get($variables, "main-font-size");
  $main-color: map.get($variables, "main-color");
  $block-margin: map.get($variables, "block-margin");
  $heading-margin: map.get($variables, "heading-margin");
  $heading-font: map.get($variables, "heading-font");
  $heading-color: map.get($variables, "heading-color");
  $heading-line-height: map.get($variables, "heading-line-height");
  $heading-letter-spacing: map.get($variables, "heading-letter-spacing");
  $heading-text-transform: map.get($variables, "heading-text-transform");
  $heading-text-shadow: map.get($variables, "heading-text-shadow");
  $heading-font-weight: map.get($variables, "heading-font-weight");
  $heading1-text-shadow: map.get($variables, "heading1-text-shadow");
  $heading1-size: map.get($variables, "heading1-size");
  $heading2-size: map.get($variables, "heading2-size");
  $heading3-size: map.get($variables, "heading3-size");
  $heading4-size: map.get($variables, "heading4-size");
  $code-font: map.get($variables, "code-font");
  $link-color: map.get($variables, "link-color");
  $link-color-hover: map.get($variables, "link-color-hover");
  $selection-background-color: map.get(
    $variables,
    "selection-background-color"
  );
  $selection-color: map.get($variables, "selection-color");

  .reveal[data-theme="#{$theme}"] {
    --r-background-color: #{$background-color};
    --r-main-font: #{$main-font};
    --r-main-font-size: #{$main-font-size};
    --r-main-color: #{$main-color};
    --r-block-margin: #{$block-margin};
    --r-heading-margin: #{$heading-margin};
    --r-heading-font: #{$heading-font};
    --r-heading-color: #{$heading-color};
    --r-heading-line-height: #{$heading-line-height};
    --r-heading-letter-spacing: #{$heading-letter-spacing};
    --r-heading-text-transform: #{$heading-text-transform};
    --r-heading-text-shadow: #{$heading-text-shadow};
    --r-heading-font-weight: #{$heading-font-weight};
    --r-heading1-text-shadow: #{$heading1-text-shadow};
    --r-heading1-size: #{$heading1-size};
    --r-heading2-size: #{$heading2-size};
    --r-heading3-size: #{$heading3-size};
    --r-heading4-size: #{$heading4-size};
    --r-code-font: #{$code-font};
    --r-link-color: #{$link-color};
    --r-link-color-hover: #{$link-color-hover};
    --r-selection-background-color: #{$selection-background-color};
    --r-selection-color: #{$selection-color};

    @content;

    @if $viewport-color {
      --r-viewport-color: #{$viewport-color};
    } @else {
      --r-viewport-color: #{$background-color};
    }
  }
}
