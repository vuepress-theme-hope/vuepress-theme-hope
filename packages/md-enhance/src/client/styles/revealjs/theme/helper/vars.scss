@use "sass:color";

// Base settings for all themes that can optionally be overridden by the super-theme

// Background of the presentation
$background-color: #2b2b2b;
$viewport-color: null;

// Primary/body text
$main-color: #eee;
$main-font: "Lato", sans-serif;
$main-font-size: 40px;

// Vertical spacing between blocks of text
$block-margin: 20px;

// Headings
$heading-margin: 0 0 $block-margin 0;
$heading-font: "League Gothic", impact, sans-serif;
$heading-color: #eee;
$heading-line-height: 1.2;
$heading-letter-spacing: normal;
$heading-text-transform: uppercase;
$heading-text-shadow: none;
$heading-font-weight: normal;
$heading1-text-shadow: $heading-text-shadow;
$heading1-size: 3.77em;
$heading2-size: 2.11em;
$heading3-size: 1.55em;
$heading4-size: 1em;
$code-font: monospace;

// Links and actions
$link-color: #13daec;
$link-color-hover: color.scale($link-color, $lightness: 20%);

// Text selection
$selection-background-color: #ff5e99;
$selection-color: #fff;
