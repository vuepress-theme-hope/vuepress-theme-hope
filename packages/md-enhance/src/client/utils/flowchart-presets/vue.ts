import base from "./base.js";

export default {
  ...base,

  // style symbol types
  symbols: {
    start: {
      class: "start-element",
      "font-color": "#fff",
      fill: "#2F495F",
      "line-width": "0px",
    },
    end: {
      class: "end-element",
      "font-color": "#fff",
      fill: "#2F495F",
      "line-width": "0px",
    },
    operation: {
      class: "operation-element",
      "font-color": "#fff",
      fill: "#00BC7D",
      "line-width": "0px",
    },
    inputoutput: {
      class: "inputoutput-element",
      "font-color": "#fff",
      fill: "#EB4D5D",
      "line-width": "0px",
    },
    subroutine: {
      class: "subroutine-element",
      "font-color": "#fff",
      fill: "#937AC4",
      "element-color": "#fff",
      "line-color": "red",
    },
    condition: {
      class: "condition-element",
      "font-color": "#fff",
      fill: "#FFB500",
      "line-width": "0px",
    },
    parallel: {
      class: "parallel-element",
      "font-color": "#fff",
      fill: "#2F495F",
      "line-width": "0px",
    },
  },
};
