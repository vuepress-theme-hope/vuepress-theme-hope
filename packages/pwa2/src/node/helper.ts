import type { ManifestOption, PWAOptions } from "../shared";

const appendBaseHelper = (base: string, link: string): string =>
  link.startsWith("/") ? link.replace(/^\//, base) : link;

export const appendBasetoManifest = (
  base: string,
  manifest: ManifestOption
): ManifestOption => {
  if (manifest.icons)
    manifest.icons = manifest.icons.map((icon) => ({
      ...icon,
      src: appendBaseHelper(base, icon.src),
    }));

  if (manifest.shortcuts)
    manifest.shortcuts = manifest.shortcuts.map((shortcut) => ({
      ...shortcut,
      icons:
        shortcut.icons?.map((icon) => ({
          ...icon,
          src: appendBaseHelper(base, icon.src),
        })) || [],
      url: appendBaseHelper(base, shortcut.url),
    }));

  if (manifest.screenshots)
    manifest.screenshots = manifest.screenshots.map((screenshot) => ({
      ...screenshot,
      src: appendBaseHelper(base, screenshot.src),
    }));

  return manifest;
};

export const appendBase = (base: string, options: PWAOptions): void => {
  if (options.favicon)
    options.favicon = appendBaseHelper(base, options.favicon);

  if (options.apple) {
    if (options.apple.icon)
      options.apple.icon = appendBaseHelper(base, options.apple.icon);

    if (options.apple.maskIcon)
      options.apple.maskIcon = appendBaseHelper(base, options.apple.maskIcon);
  }

  if (options.msTile && options.msTile.image)
    options.msTile.image = appendBaseHelper(base, options.msTile.image);

  if (options.manifest)
    options.manifest = appendBasetoManifest(base, options.manifest);
};
