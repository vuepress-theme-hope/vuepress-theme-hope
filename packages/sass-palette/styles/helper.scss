@use "sass:color";
@use "sass:list";
@use "sass:map";
@use "sass:meta";
@use "sass:string";

@function is-upper-case($char) {
  @return string.to-lower-case($char) != $char;
}

@function is-lower-case($char) {
  @return string.to-upper-case($char) != $char;
}

@function camel-to-kebab($string) {
  $length: string.length($string);
  $property: "--";

  @for $i from 1 through $length {
    $char: string.slice($string, $i, $i);

    $property: "#{$property}#{if(is-upper-case($char), "-#{string.to-lower-case(
        $char
      )}",$char)}";
  }

  @return $property;
}

@function color-is-legal($value) {
  // single color
  @if meta.type-of($value) == color {
    @return true;
  }

  // itâ€™s a valid map color config
  @if (meta.type-of($value) == map) {
    @if (meta.type-of(map.get($value, "light")) == color) and
      (meta.type-of(map.get($value, "dark")) == color)
    {
      @return true;
    }

    @if (is-legal-config(map.get($value, "base"))) {
      $scale: map.get($value, "scale");

      @if (meta.type-of($scale) == map) {
        @each $name, $amount in $scale {
          @if (meta.type-of($amount) != number) {
            @debug "Unexpected value: getting #{$value}";

            @return false;
          }
        }

        @return true;
      }
    }
  }

  @warn "Unexpected value: getting #{$value}";

  @return false;
}

@function get-color($color, $isDark: false) {
  @if meta.type-of($color) == color {
    @return $color;
  }

  $lightColor: map.get($color, "light");
  $darkColor: map.get($color, "dark");

  @if $lightColor and $darkColor {
    @return if($isDark, $darkColor, $lightColor);
  }

  @warn "Unexpected color: getting #{$color}";

  @return transparent;
}

@mixin inject-color($key, $config, $dark-selector: "") {
  @if meta.type-of($config) == color {
    :root {
      #{$key}: #{$config};
    }
  } @else {
    $lightColor: map.get($config, "light");
    $darkColor: map.get($config, "dark");

    @if $lightColor and $darkColor {
      :root {
        #{$key}: #{$lightColor};
      }

      @if $lightColor != $darkColor and $dark-selector {
        #{$dark-selector} {
          #{$key}: #{$darkColor};
        }
      }
    }
  }
}
