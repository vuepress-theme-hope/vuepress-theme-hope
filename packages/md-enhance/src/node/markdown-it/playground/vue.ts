import { deepAssign, entries, fromEntries } from "@vuepress/helper";

import type {
  PlaygroundData,
  PlaygroundOptions,
  VuePresetPlaygroundOptions,
} from "../../typings/index.js";

const VUE_SUPPORTED_EXTENSIONS = [
  "html",
  "js",
  "ts",
  "vue",
  "jsx",
  "tsx",
  "json",
];

const DEFAULT_VUE_CDN = "https://sfc.vuejs.org/vue.runtime.esm-browser.js";
const DEFAULT_VUE_SERVER_RENDERER_CDN =
  "https://sfc.vuejs.org/server-renderer.esm-browser.js";

export const getVuePlaygroundPreset = (
  options: VuePresetPlaygroundOptions = {},
): PlaygroundOptions => ({
  name: "playground#vue",
  propsGetter: (playgroundData: PlaygroundData): Record<string, string> => {
    const { title = "", files, settings: localSettings } = playgroundData;
    const settings = {
      // Defaults
      service: "https://sfc.vuejs.org/",
      dev: false,
      ssr: false,
      ...options,
      ...localSettings,
    };

    const fileInfo = fromEntries(
      entries(files)
        .filter(([, { ext }]) => VUE_SUPPORTED_EXTENSIONS.includes(ext))
        .map(([key, { content }]) => {
          if (key === "import-map.json") {
            const importMap = <
              {
                imports: Record<string, string>;
                scopes?: Record<string, Record<string, string>>;
              }
            >JSON.parse(content);

            return [
              key,
              JSON.stringify(
                deepAssign(
                  {
                    imports: {
                      // Insure vue exists
                      vue: DEFAULT_VUE_CDN,
                      // Insure vue/server-renderer exists
                      ...(settings.ssr
                        ? {
                            // eslint-disable-next-line @typescript-eslint/naming-convention
                            "vue/server-renderer":
                              DEFAULT_VUE_SERVER_RENDERER_CDN,
                          }
                        : {}),
                    },
                  },
                  importMap,
                ),
                null,
                2,
              ),
            ];
          }

          return [key, content];
        }),
    );

    if (settings.ssr && !fileInfo["import-map.json"])
      fileInfo["import-map.json"] = JSON.stringify(
        {
          imports: {
            vue: DEFAULT_VUE_CDN,
            // eslint-disable-next-line @typescript-eslint/naming-convention
            "vue/server-renderer": DEFAULT_VUE_SERVER_RENDERER_CDN,
          },
        },
        null,
        2,
      );

    return {
      title,
      link: encodeURIComponent(
        `${settings.service}#${
          // Dev flag
          settings.dev ? "__DEV__" : ""
        }${
          // Ssr flag
          settings.ssr ? "__SSR__" : ""
        }${
          // Code base64
          Buffer.from(JSON.stringify(fileInfo)).toString("base64")
        }`,
      ),
    };
  },
});
