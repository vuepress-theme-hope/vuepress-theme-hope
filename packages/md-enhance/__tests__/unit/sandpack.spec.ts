/* eslint-disable @typescript-eslint/naming-convention */
import { decodeData } from "@vuepress/helper";
import MarkdownIt from "markdown-it";
import { describe, expect, it } from "vitest";

import { sandpack } from "../../src/node/markdown-it/sandpack/index.js";

const decodeFiles = (content: string): Record<string, string> =>
  JSON.parse(decodeData(content)) as Record<string, string>;

const getTemplate = (renderResult: string): string | null => {
  const result = renderResult.match(/template="(.*?)"/s);

  if (!result) return null;

  return result[1];
};

const getFiles = (renderResult: string): Record<string, string> | null => {
  const result = renderResult.match(/files="(.*?)"/s);

  if (!result) return null;

  return decodeFiles(result[1]);
};

const getOptions = (renderResult: string): Record<string, unknown> | null => {
  const result = renderResult.match(/options="(.*?)"/s);

  if (!result) return null;

  return JSON.parse(decodeData(result[1])) as Record<string, unknown>;
};

const getCustomSetup = (
  renderResult: string,
): Record<string, unknown> | null => {
  const result = renderResult.match(/customSetup="(.*?)"/s);

  if (!result) return null;

  return JSON.parse(decodeData(result[1])) as Record<string, unknown>;
};

describe("Sandpack", () => {
  const markdownIt = MarkdownIt({ linkify: true }).use(sandpack);

  it("Should resolve sandpack info", () => {
    const result = markdownIt.render(
      `
::: sandpack#vue Sandpack demo

@file App.vue
\`\`\`vue
<script setup>
import { ref } from 'vue'
const msg = ref('Hello World!')
</script>
<template>
  <h1>{{ msg }}</h1>
  <input v-model="msg" />
</template>
\`\`\`

@file Comp.vue

\`\`\`vue
<template>
  <div>Comp</div>
</template>
\`\`\`

:::
`,
      {},
    );

    expect(result).toMatchSnapshot();

    const template = getTemplate(result);
    const files = getFiles(result);
    const options = getOptions(result);
    const customSetup = getCustomSetup(result);

    expect(template).toEqual("vue");

    expect(files).toEqual({
      "App.vue": `\
<script setup>
import { ref } from 'vue'
const msg = ref('Hello World!')
</script>
<template>
  <h1>{{ msg }}</h1>
  <input v-model="msg" />
</template>
`,
      "Comp.vue": `\
<template>
  <div>Comp</div>
</template>
`,
    });

    expect(options).toEqual({});
    expect(customSetup).toEqual({});
  });

  it("Should resolve sandpack info with options", () => {
    const result = markdownIt.render(
      `
::: sandpack Sandpack demo with options

@file App.vue

\`\`\`vue
<script setup>
import { ref } from 'vue'
const msg = ref('Hello World!')
</script>
<template>
  <h1>{{ msg }}</h1>
  <input v-model="msg" />
</template>
\`\`\`

@file Comp.vue

\`\`\`vue
<template>
  <div>Comp</div>
</template>
\`\`\`

@options

\`\`\`js
{
  readOnly: true,
  showReadOnly: false
}
\`\`\`

:::
`,
      {},
    );

    expect(result).toMatchSnapshot();

    const template = getTemplate(result);
    const file = getFiles(result);
    const options = getOptions(result);
    const customSetup = getCustomSetup(result);

    expect(template).toEqual(null);

    expect(file).toEqual({
      "App.vue": `\
<script setup>
import { ref } from 'vue'
const msg = ref('Hello World!')
</script>
<template>
  <h1>{{ msg }}</h1>
  <input v-model="msg" />
</template>
`,
      "Comp.vue": `\
<template>
  <div>Comp</div>
</template>
`,
    });
    expect(options).toEqual({
      readOnly: true,
      showReadOnly: false,
    });

    expect(customSetup).toEqual({});
  });

  it("Should resolve sandpack info with customSetup", () => {
    const result = markdownIt.render(
      `
::: sandpack#vue Sandpack demo with customSetup

@file App.vue

\`\`\`vue
<script setup>
import { ref } from 'vue'
const msg = ref('Hello World!')
</script>
<template>
  <h1>{{ msg }}</h1>
  <input v-model="msg" />
</template>
\`\`\`

@options

\`\`\`js
{
  readOnly: true,
  showReadOnly: false
}
\`\`\`

@setup

\`\`\`js
{
  entry: "/index.js",
}
\`\`\`

:::
`,
      {},
    );

    expect(result).toMatchSnapshot();

    const template = getTemplate(result);
    const file = getFiles(result);
    const options = getOptions(result);
    const customSetup = getCustomSetup(result);

    expect(template).toEqual("vue");

    expect(file).toEqual({
      "App.vue": `\
<script setup>
import { ref } from 'vue'
const msg = ref('Hello World!')
</script>
<template>
  <h1>{{ msg }}</h1>
  <input v-model="msg" />
</template>
`,
    });

    expect(options).toEqual({
      readOnly: true,
      showReadOnly: false,
    });

    expect(customSetup).toEqual({
      entry: "/index.js",
    });
  });

  it("Should resolve sandpack info with file attrs", () => {
    const result = markdownIt.render(
      `
::: sandpack Sandpack demo with options

@file /src/App.vue [readOnly hidden]

\`\`\`vue
<script setup>
import { ref } from 'vue'
const msg = ref('Hello World!')
</script>
<template>
  <h1>{{ msg }}</h1>
  <input v-model="msg" />
</template>
\`\`\`

@file /src/Comp.vue [active]

\`\`\`vue
<template>
  <div>Comp</div>
</template>
\`\`\`

@file Comp2.vue \\[hidden]

\`\`\`vue
<template>
  <div>Comp2</div>
</template>
\`\`\`

:::
`,
      {},
    );

    expect(result).toMatchSnapshot();

    const template = getTemplate(result);
    const file = getFiles(result);
    const options = getOptions(result);
    const customSetup = getCustomSetup(result);

    expect(template).toEqual(null);

    expect(file).toEqual({
      "/src/App.vue": {
        code: `\
<script setup>
import { ref } from 'vue'
const msg = ref('Hello World!')
</script>
<template>
  <h1>{{ msg }}</h1>
  <input v-model="msg" />
</template>
`,
        active: false,
        hidden: true,
        readOnly: true,
      },
      "/src/Comp.vue": {
        code: `\
<template>
  <div>Comp</div>
</template>
`,
        active: true,
        hidden: false,
        readOnly: false,
      },
      "Comp2.vue": `\
<template>
  <div>Comp2</div>
</template>
`,
    });
    expect(options).toEqual({});

    expect(customSetup).toEqual({});
  });

  it("Should resolve sandpack info with file attrs and customSetup", () => {
    const result = markdownIt.render(
      `
::: sandpack Sandpack demo with file attrs and customSetup [rtl theme=dark]

@file /src/App.vue [readOnly]

\`\`\`vue
<script setup>
import { ref } from "vue";
import Comp from "./Comp.vue";

const msg = ref("Hello Playground!");
</script>

<template>
  <h1>{{ msg }}</h1>
  <input v-model="msg" />
  <Comp />
</template>
\`\`\`

@file /src/Comp.vue [active]

\`\`\`vue
<script setup>
import { useBattery } from "@vueuse/core";
import { ref } from "vue";

const { charging, level } = useBattery();
</script>

<template>
  <h1>Battery status</h1>
  <p>Charging: {{ charging }}</p>
  <p>Level: {{ level * 100 }}%</p>
</template>
\`\`\`

@setup

\`\`\`js
{
  dependencies: {
    "@vueuse/core": "latest",
    "@vueuse/shared": "latest",
    "vue-demi": "latest",
  }
}
\`\`\`

:::
`,
      {},
    );

    expect(result).toMatchSnapshot();
    expect(result).includes("rtl");
    expect(result).includes('theme="dark"');

    const template = getTemplate(result);
    const file = getFiles(result);
    const options = getOptions(result);
    const customSetup = getCustomSetup(result);

    expect(template).toEqual(null);

    expect(file).toEqual({
      "/src/App.vue": {
        code: `\
<script setup>
import { ref } from "vue";
import Comp from "./Comp.vue";

const msg = ref("Hello Playground!");
</script>

<template>
  <h1>{{ msg }}</h1>
  <input v-model="msg" />
  <Comp />
</template>
`,
        active: false,
        hidden: false,
        readOnly: true,
      },
      "/src/Comp.vue": {
        code: `\
<script setup>
import { useBattery } from "@vueuse/core";
import { ref } from "vue";

const { charging, level } = useBattery();
</script>

<template>
  <h1>Battery status</h1>
  <p>Charging: {{ charging }}</p>
  <p>Level: {{ level * 100 }}%</p>
</template>
`,
        active: true,
        hidden: false,
        readOnly: false,
      },
    });
    expect(options).toEqual({});

    expect(customSetup).toEqual({
      dependencies: {
        "@vueuse/core": "latest",
        "@vueuse/shared": "latest",
        "vue-demi": "latest",
      },
    });
  });
});
