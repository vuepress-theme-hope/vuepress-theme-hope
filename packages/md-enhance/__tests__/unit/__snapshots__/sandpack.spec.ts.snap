// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Sandpack > Should resolve sandpack info 1`] = `
"<SandPack title="Sandpack%20demo" template="vue" files="eJxdjrsKAjEQRX9lTBMFNdhKXBAb/8AmjexmNZDHkMxuE/LvJiuI2t2ZO+cwmZ0R9/Ok2ZHJ1EeDBEnThJ3yxmGIBBmiHqHAGIMDXk+58n3wicClB5xau+ZXbW2AW4h2WPGN8lK8ZVUjSTu0d9I1A8jnoct5QUuRok7L1nicCOadC4O2J8VqrxiIhosvnm3ZJbjPw7/mwcxda6Vo6Z8sLze/UE4=" options="eJyrrgUAAXUA+Q==" customSetup="eJyrrgUAAXUA+Q==">
</SandPack>
"
`;

exports[`Sandpack > Should resolve sandpack info with customSetup 1`] = `
"<SandPack title="Sandpack%20demo%20with%20customSetup" template="vue" files="eJxNjLsOwiAUhl/lyIIOSlwNbeLmG7iwmPZUSbicwKEL4d2ldXH7b99fxZ3oshYUN6HzlCwxZORCownWU0wMFRIu0GBJ0YPsU2nCFENm8PkNw9Ye5QOdi/CMyc0HeTJBq99Zv9GMntyLsWsA/bmOte5oa1p1t6c2UGFYzz7O6AYjem8EqA1Xf7xoX0y4PKI=" options="eJyrVipKTUzxz8upVLIqKSpN1VEqzsgvD4KLpSXmFKfWAgAN5g4J" customSetup="eJyrVkrNKymqVLJS0s/MS0mt0MsqVqoFAE+4Bz8=">
</SandPack>
"
`;

exports[`Sandpack > Should resolve sandpack info with file attrs 1`] = `
"<SandPack title="Sandpack%20demo%20with%20options" files="eJx1j71OBCEUhV/lSjOajBK33LCTGBs7SxuaCTAuCX8BZhJDeHcvg0aNuxX3h3O+cwuhKQr6FMLDtipyLER4iS9hSUQdMiSV1zBxp23wMUOBqBaosERvYUDJwJ3wLmWw6R1ObXs7vChjPLz5aOTNcMcdo90MbVhWNpg5K6wB2PlxKmWX1soodvtUu7Bm2O4tRjEnTnDPCdAmp7/0ZCSzyHrDuMtskhrJWUupHDnmuGIX1SxfnfnofR37qc/e/rv1byipt6n9YrRV16Gd8s38ivAD3QdIbVaHTrxGOlxG1U+b7o11" options="eJyrrgUAAXUA+Q==" customSetup="eJyrrgUAAXUA+Q==">
</SandPack>
"
`;

exports[`Sandpack > Should resolve sandpack info with file attrs and customSetup 1`] = `
"<SandPack rtl theme="dark" title="Sandpack%20demo%20with%20file%20attrs%20and%20customSetup%20" files="eJyNkE1OwzAQha8yWEICFOp2W9IK6IYFEhzAGyuZppEc2/JPpCrK3Rk7pC2Loq6SmXl+b+YbGPeu4m/WLvqIbD2wytT0ZaWvXGsDeAzRboVuO2tcgAEc7mGEvTMdCEZvBHs5TXems/NowVOVXLNC6MpoH6DzDWySyYNgH6iUgW8lj40zUdd3gj2StORTNqVSEbCzSgakCqA8rLbDkE3GseRU5W6rbQzQP3e0u9oIRnPBgE/DvFT6L/mFFyuYrELb0617qTwW7NDWNepT6VDWX1od2Tq4iGMxgZpvuoVU9PguQ0B3PAN7pbfU55Vxf8hd4TpTG6A6SNe0uilAYY+KpJuLgIcbuM27+CBD9Gd4drv79V4DsZ2DMuB0T5Z8ptA8n+KfYLVckuR+0lxDm9D9QzY3xvEH49nd3w==" options="eJyrrgUAAXUA+Q==" customSetup="eJyrVkpJLUjNS0nNS85MLVayqlZyKCtNLS1O1U/OL0pVslLKSSxJLS5R0oGLF2ckFqWmIMsAJXRTUnMzEWK1tQBugx7v">
</SandPack>
"
`;

exports[`Sandpack > Should resolve sandpack info with options 1`] = `
"<SandPack title="Sandpack%20demo%20with%20options" files="eJxdjrsKAjEQRX9lTBMFNdhKXBAb/8AmjexmNZDHkMxuE/LvJiuI2t2ZO+cwmZ0R9/Ok2ZHJ1EeDBEnThJ3yxmGIBBmiHqHAGIMDXk+58n3wicClB5xau+ZXbW2AW4h2WPGN8lK8ZVUjSTu0d9I1A8jnoct5QUuRok7L1nicCOadC4O2J8VqrxiIhosvnm3ZJbjPw7/mwcxda6Vo6Z8sLze/UE4=" options="eJyrVipKTUzxz8upVLIqKSpN1VEqzsgvD4KLpSXmFKfWAgAN5g4J" customSetup="eJyrrgUAAXUA+Q==">
</SandPack>
"
`;
